pre-aio-node-with-zuul-ref.sh

```diff
@@ -0,0 +1,157 @@
+
+IFS='^' read -r -a ZUUL_PROJ_CHANGE <<< "$ZUUL_CHANGES"
+related_proj=()
+for proj in ${ZUUL_PROJ_CHANGE[@]}
+do
+    proj=${proj%%:*}
+    proj=${proj##*/}
+    related_proj+=($proj)
+done
+unique_proj_list=$(echo "${related_proj[@]}" | tr ' ' '\n' | nl | sort -u -k2 | sort -n | cut -f2-)
+
+echo ${DEST}
+cd ${DEST}
+
+for dir in ${DEST}/*
+do
+    if [ -d "${dir}" ] ; then
+        echo "check ${dir}"
+        cd ${dir}
+        if [ -d .git ]; then
+            echo "update repository ${dir}"
+            git checkout master
+            git pull --ff origin master
+
+            pkg_flag=${dir##*/}
+            echo $pkg_flag
+            if [[ " ${unique_proj_list[@]} " =~ " ${pkg_flag} " ]]; then
+                echo "merge repository ${dir}"
+                git branch -D $TMP_MERGE_BRANCH
+                git checkout master -B $TMP_MERGE_BRANCH
+                git fetch ${FETCH_URL}/openstack/${pkg_flag} $refspec
+                git merge FETCH_HEAD -m "temp branch for vgpu patches merge"
+            fi
+        fi;
+    fi
+done
+
+first_vgpu_type=$(on_xenserver << END_OF_REQ_VGPU_TYPE
+vgpu_type_list=\$(xe vgpu-type-list --minimal)
+vgpu_type_list=\${vgpu_type_list//,/ }
+for id in \$vgpu_type_list
+do
+    name=\$(xe vgpu-type-param-get uuid=\$id param-name="model-name")
+    if [ "\$name" != "passthrough" ]
+    then
+        echo \$name
+        exit
+    fi
+done
+END_OF_REQ_VGPU_TYPE
+)
+
+cd $DEVSTACK_TOOL/
+./stack.sh
+echo "[devices]" >> ${NOVA_CONF_PATH}/nova.conf
+echo "enabled_vgpu_types = $first_vgpu_type" >> ${NOVA_CONF_PATH}/nova.conf
+sudo systemctl restart devstack@n-cpu.service
+
+
+cd $TEMPEST_DIR
+# TODO: run vgpu tempest
+
+source $DEVSTACK_TOOL/openrc admin
+nova flavor-key 1 set resources:VGPU=1
+image_id=$(glance image-list | grep "$SAMPLE_IMAGE" | awk '{print $2}')
+glance image-update  --property vm_mode=hvm $image_id
+
+source $DEVSTACK_TOOL/openrc demo demo
+prv_net=$(openstack network list | grep "private" | awk '{print $2}')
+
+nova boot --image $SAMPLE_IMAGE --flavor 1 --nic net-id=$prv_net $VM_NAME
+
+nova_vm_id=$(nova show $VM_NAME | grep -w 'id' | awk '{print $4}')
+
+i=0
+while :
+do
+    sleep 5
+    i=`expr $i + 1`
+    vm_state=$(nova list  --name $VM_NAME --fields status | grep ACTIVE)
+    if [ -n "$vm_state" ]; then
+        break
+    elif [ $i -gt 20 ]; then
+        echo "VGPU create failed"
+        sleep 5
+        exit
+    fi
+done
+
+result=$(on_xenserver << END_OF_VGPU_CONFIRM
+set -x
+vgpu_list=\$(xe vgpu-list --minimal)
+vgpu_list=\${vgpu_list//,/ }
+for vgpu_id in \$vgpu_list
+do
+    ret=\$(xe vgpu-param-list uuid=\$vgpu_id | grep $nova_vm_id)
+    echo \$ret
+done
+set +x
+END_OF_VGPU_CONFIRM
+)
+
+if [ -n "$result" ]; then
+    echo "VGPU create successfully"
+else
+    echo "VGPU create failed"
+fi
+
+# Post proccess
+echo $keep_env
+if [ $keep_env != "true" ]; then
+    cd $DEVSTACK_TOOL/
+    ./clean.sh
+fi
+sed -i '/DEST=${DEST}/d' ${DEVSTACK_TOOL}/local.conf
```

Comment:
This is not needed. When adding this option, we can't depends on it's deleted at here. Instead we should check if that's already configured correctly before adding it.

Reply:
OK, will modify it